Implement histogram statistics decoder

fpetkovski opened the pull request:
This PR speeds up `histogram_count` and `histogram_sum` functions on native histograms. The idea is to use a separate decoder which can be used by the engine to skip histogram buckets when decoding histogram objects. This should help with reducing allocations and make the sum aggregation faster it will only be done on primitive values excluding buckets and spans.

Current benchmark results
```
prometheus $ benchstat main.out new.out                                               
name                                                          old time/op    new time/op    delta
NativeHistograms/sum-11                                          404ms ± 6%     391ms ± 1%     ~     (p=0.222 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            500ms ± 8%     492ms ± 8%     ~     (p=0.548 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             803ms ± 1%     815ms ± 3%     ~     (p=0.095 n=5+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     477ms ± 4%     318ms ± 5%  -33.29%  (p=0.008 n=5+5)
NativeHistograms/histogram_count_with_long_rate_interval-11      818ms ± 4%     582ms ± 1%  -28.77%  (p=0.008 n=5+5)

name                                                          old alloc/op   new alloc/op   delta
NativeHistograms/sum-11                                          413MB ± 0%     414MB ± 0%     ~     (p=0.841 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            213MB ± 0%     213MB ± 0%     ~     (p=0.095 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             314MB ± 0%     314MB ± 0%     ~     (p=0.841 n=5+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     213MB ± 0%     143MB ± 0%  -32.94%  (p=0.016 n=5+4)
NativeHistograms/histogram_count_with_long_rate_interval-11      321MB ± 2%     202MB ± 6%  -37.15%  (p=0.008 n=5+5)

name                                                          old allocs/op  new allocs/op  delta
NativeHistograms/sum-11                                          5.16M ± 0%     5.16M ± 0%     ~     (p=1.000 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            4.61M ± 0%     4.61M ± 0%     ~     (p=0.381 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             6.12M ± 0%     6.12M ± 0%     ~     (p=1.000 n=5+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     4.61M ± 0%     1.66M ± 0%  -63.99%  (p=0.008 n=5+5)
NativeHistograms/histogram_count_with_long_rate_interval-11      6.14M ± 0%     2.01M ± 3%  -67.36%  (p=0.008 n=5+5)
```

<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

fpetkovski commented:
I've been tinkering with this approach a bit and I wonder if it could interfere with various mixed types checks, as well as with histogram counter resets. I will investigate another solution which returns a partially filled histogram object so that we can still go through the same `histogramRate` path.

fpetkovski commented:
Looking at profiles, I see `CopyToSchema` taking a lot of time. Do we need to decode/fill the schema field if we're only operating on Count/Sum values @krajorama ?
![image](https://github.com/prometheus/prometheus/assets/1286231/9af27f89-3539-49dc-b233-a665b1a968da)

beorn7 commented:
Generally, this looks very nice.

However, there is a nasty edge case: In rare cases, a counter reset could not be seen in the count but only in individual buckets. We do detect those cases on ingestion and set the counter reset hint accordingly, so generally, this should make it through the special decoder here via the counter reset hint. However (edge case of an edge case of an edge case…), the following scenario could happen:
- A chunk gets cut just because the previous chunk was full (and not because of a counter reset). In that case, we mark the counter reset hint as "unknown" because we cannot know if the previous chunk gets deleted. The idea is that we do manual counter reset detection in that case, just to play safe.
- If the previous chunk gets really deleted, and if there is really a gap between the pre-previous chunk and the current chunk, and if that counter reset isn't represented in the count, but in (some of) the buckets, then this PR creates a problem (a different result than the one we get with the full decoding).

Maybe we can solve that cleanly by doing the counter reset detection directly in the iterator (and set the counter reset accordingly) in cases where the "unknown" hint is coming from TSDB.

Or we can just wing it and claim that this case is so outrageously unlikely that we don't need to handle it.

beorn7 commented on tsdb/chunkenc/histogram.go:
```diff
@@ -1121,3 +1121,71 @@ func resize[T any](items []T, n int) []T {
+			ZeroCount:        f.hReader.ZeroCount,
+			Sum:              f.hReader.Sum,
+			ZeroThreshold:    f.hReader.ZeroThreshold,
+			Schema:           f.hReader.Schema,
```
Since the schema is only for buckets, I think we can set this to any number without doing harm. If we set it consistently to the same number (e.g. 0), we avoid the CopyToSchema step.

fpetkovski replied:
I removed both the zero threshold and schema fields.

beorn7 commented on promql/engine.go:
```diff
@@ -3313,8 +3320,45 @@ func setOffsetForAtModifier(evalTime int64, expr parser.Expr) {
 	})
 }
 
+func detectHistogramStatsDecoding(expr parser.Expr) {
```
Let's have a doc comment here explaining what this does.

fpetkovski replied:
Added a comment.

MichaHoffmann commented on promql/engine.go:
```diff
@@ -3313,8 +3320,45 @@ func setOffsetForAtModifier(evalTime int64, expr parser.Expr) {
 
+func detectHistogramStatsDecoding(expr parser.Expr) {
+	var readHistogramStats bool
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
```
i think i would feel safer if we would only inspect the leaf nodes and then check the path if the last function up the tree was a histogram_sum or _count. That way its stateless and independent of the traversal.

fpetkovski replied:
Updated, ptal again.

MichaHoffmann replied:
lgtm

fpetkovski commented:
We currently don't have a function to detect resets for integer histograms. Since PromQL always works with float histograms, I might need to add the same functionality just operating on integer buckets and spans.

beorn7 reviewed (changes requested):
Just some comment nits.

beorn7 commented:
> Let me take a look at how we can do the detection transparently in an efficient manner. In the worst case we will just store the last decoded histogram as another field and use it for detection when the hint is unknown.

Any ideas so far?

It would actually be quite cool if we could detect counter resets in the iterator already. Then we could remove it entirely in the PromQL layer (for native histograms, at least).

If we can somehow safely assess if a chunk got deleted before the current chunk, we could avoid the manual counter reset detection even more. A heuristic would be if there is a gap larger than the typical scrape interval, but that's not strictly correct. I have a vague memory that we never delete chunks from the middle of a block, so that might narrow down the problem.

Just brainstorming here…

I'm almost ready to believe that the above mentioned edge case of an edge case of an edge case is rare enough that we can ignore it (but document it, just in case it becomes relevant after all).

beorn7 commented:
Another thought that crossed my mind: What about overlapping blocks and (soon) out-of-order handling?
This might create a lot more cases where we set the counter reset hint to "unknown". Maybe we should simply skip the optimization in those cases. (Again, I have to apologize that I just type out thoughts crossing my mind, based on vague memories from the past, without having done a proper research into the matter…)

fpetkovski commented:
I've been busy moving apartments this week and did not have time to work in this further. I'll have my time back next week and will continue looking into the reset. 

My opinion is that we should handle the reset in the iterator by keeping track of the last decoded histogram. This way the code can also remain future proof in case we introduce more edge cases down the line.

beorn7 commented:
Good luck with the move. Looking forward to further explorations into this issue.

fpetkovski commented:
@beorn7 I've addressed all comments and we now detect the reset in the iterator itself. I also moved iterator to the promql package since this is the only place where it is used. It also relies on public methods so we don't need access to encoding internals from the `chunkenc` package.

fpetkovski commented on promql/engine.go:
```diff
@@ -3313,8 +3320,50 @@ func setOffsetForAtModifier(evalTime int64, expr parser.Expr) {
+// detectHistogramStatsDecoding modifies the expression by setting the
+// SkipHistogramBuckets field in those vector selectors for which it is safe to
+// return only histogram statistics (sum and count), excluding histogram spans
+// and buckets. The function can be treated as an optimization and does is not
```
Yeah this was spurious.

beorn7 commented:
Thank you very much. This is now very close to the top of my review queue, but I won't get to it this week. Sorry for that.

beorn7 commented on promql/histogram_stats_iterator.go:
```diff
@@ -0,0 +1,134 @@
+	lastFH   *histogram.FloatHistogram
+}
+
+func NewHistogramStatsIterator(it chunkenc.Iterator) chunkenc.Iterator {
```
How about some doc comments (here and as applicable for the methods below) what this is for and how it works?

beorn7 commented on promql/histogram_stats_iterator.go:
```diff
@@ -0,0 +1,134 @@
+type histogramStatsIterator struct {
+	chunkenc.Iterator
+
+	hReader *histogram.Histogram
```
"Reader" somehow feels wrong as a name (in both `hReader` and `fhReader`). Maybe `currentH` and `currentFH`? Or just use `h` and `fh` (not saying anything specific will avoid assumptions…)?

beorn7 commented on promql/histogram_stats_iterator.go:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+func (f *histogramStatsIterator) AtHistogram(h *histogram.Histogram) (int64, *histogram.Histogram) {
```
I guess the doc comment here and below for `AtFloatHistogram` should mention that the rest detection only works if you don't switch between `AtHistogram` and `AtFloatHistogram`. (PromQL only ever uses the latter, so that should be fine in practice.)

beorn7 reviewed (changes requested):
Very neat, thank you very much. I believe (and hope) this works as we expect. :o)

One additional idea for testing: It should be quite easy to write tests within the PromqL testing framework (in https://github.com/prometheus/prometheus/blob/main/promql/promqltest/testdata/native_histograms.test ) that use PromQL functions that use the new stats iterator under the hood. We could even construct counter resets there that are only detectable in buckets, not in the count (but I think it will be too hard to provoke a new chunk being cut at the same time so I don't think we'll exercise the new counter reset code path, because the testing framework probably ingests normally and will store the counter reset in the chunk header already).

beorn7 commented:
How does the benchmark look with the added storing of the last seen histogram?

fpetkovski commented:
Thanks for the review, I addressed all comments. Benchmarks still look good on the latest commit:

```
name                                                          old time/op    new time/op    delta
NativeHistograms/sum-11                                          396ms ± 2%     400ms ± 1%     ~     (p=0.310 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            480ms ± 2%     481ms ± 2%     ~     (p=0.841 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             800ms ± 1%     809ms ± 1%     ~     (p=0.095 n=5+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     473ms ± 0%     302ms ± 0%  -36.14%  (p=0.008 n=5+5)
NativeHistograms/histogram_count_with_long_rate_interval-11      813ms ± 5%     513ms ± 1%  -36.89%  (p=0.008 n=5+5)

name                                                          old alloc/op   new alloc/op   delta
NativeHistograms/sum-11                                          414MB ± 0%     413MB ± 0%     ~     (p=0.222 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            213MB ± 0%     213MB ± 0%     ~     (p=0.222 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             314MB ± 0%     314MB ± 0%     ~     (p=0.690 n=5+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     213MB ± 0%     137MB ± 0%  -35.80%  (p=0.008 n=5+5)
NativeHistograms/histogram_count_with_long_rate_interval-11      330MB ± 0%     140MB ± 8%  -57.48%  (p=0.016 n=4+5)

name                                                          old allocs/op  new allocs/op  delta
NativeHistograms/sum-11                                          5.16M ± 0%     5.16M ± 0%     ~     (p=0.095 n=5+5)
NativeHistograms/sum_rate_with_short_rate_interval-11            4.61M ± 0%     4.61M ± 0%     ~     (p=0.690 n=5+5)
NativeHistograms/sum_rate_with_long_rate_interval-11             6.12M ± 0%     6.12M ± 0%     ~     (p=0.603 n=4+5)
NativeHistograms/histogram_count_with_short_rate_interval-11     4.61M ± 0%     1.61M ± 0%  -65.04%  (p=0.008 n=5+5)
NativeHistograms/histogram_count_with_long_rate_interval-11      6.16M ± 1%     1.52M ± 3%  -75.23%  (p=0.008 n=5+5)
```

beorn7 reviewed (approved):
Thank you very much, let's ship it!

beorn7 commented:
I hope it's OK if I squash the commits into one…
